modin/core/storage_formats/pandas/merge.py

```diff
@@ -0,0 +1,299 @@
+            # it's fine too, we can also decide that by columns, which tend to be already
+            # materialized quite often compared to the indexes.
+            keep_index = False
+            if left._modin_frame.has_materialized_index:
+                keep_index = should_keep_index(left, right)
+            else:
+                # Have to trigger columns materialization. Hope they're already available at this point.
+                if left_on is not None and right_on is not None:
+                    keep_index = any(
+                        o not in right.columns
+                        and o in left_on
+                        and o not in left.columns
+                        for o in right_on
+                    )
+                elif on is not None:
+                    keep_index = any(
+                        o not in right.columns and o not in left.columns for o in on
+                    )
+
+            if sort:
+                if left_on is not None and right_on is not None:
+                    new_left = (
+                        new_left.sort_index(axis=0, level=left_on + right_on)
+                        if keep_index
+                        else new_left.sort_rows_by_column_values(left_on + right_on)
+                    )
+                elif on is not None:
+                    new_left = (
+                        new_left.sort_index(axis=0, level=on)
+                        if keep_index
+                        else new_left.sort_rows_by_column_values(on)
+                    )
+
+            return (
+                new_left.reset_index(drop=True)
+                if not keep_index and (kwargs["how"] != "left" or sort)
+                else new_left
+            )
+        else:
+            return left.default_to_pandas(pandas.DataFrame.merge, right, **kwargs)
+
+    @classmethod
+    def _compute_result_metadata(cls, left, right, on, left_on, right_on, suffixes):
+        """
+        Compute columns and dtypes metadata for the result of merge if possible.
+
+        Parameters
+        ----------
+        left : PandasQueryCompiler
+        right : PandasQueryCompiler
+        on : label, list of labels or None
+            `on` argument that was passed to ``pandas.merge()``.
+        left_on : label, list of labels or None
+            `left_on` argument that was passed to ``pandas.merge()``.
+        right_on : label, list of labels or None
+            `right_on` argument that was passed to ``pandas.merge()``.
+        suffixes : list of strings
+            `suffixes` argument that was passed to ``pandas.merge()``.
+
+        Returns
+        -------
+        new_columns : pandas.Index or None
+            Columns for the result of merge. ``None`` if not enought metadata to compute.
+        new_dtypes : ModinDtypes or None
+            Dtypes for the result of merge. ``None`` if not enought metadata to compute.
+        """
+        new_columns = None
```

Comment:
this logic was copied without any changes from `PandasQueryCompiler.merge` and was placed in a separate method to be reused by range-partitioning impl